src/library/scala/collection/immutable/ListMap.scala

```diff
@@ -275,24 +294,50 @@ private[immutable] final class ListMapBuilder[K, V] extends mutable.ReusableBuil
     }
     this
   }
+  override def addAll(xs: IterableOnce[(K, V)]): this.type = {
+    if (isAliased) {
+      super.addAll(xs)
+    } else if (underlying.nonEmpty) {
+      xs match {
+        case m: collection.Map[K, V] =>
+          // if it is a map, then its keys will not collide with themselves.
```

Comment:
You know, this isn't _entirely_ true if someone uses an `OrderedMap` where the ordering is more precise than equality.  However, I think that's an abuse of the API so people shouldn't be surprised if some random things break.

Reply:
The old collections guide is silent on this matter. I put this into the new map/set rules (https://github.com/scala/scala/blob/2.13.x/test/junit/scala/collection/SetMapRulesTest.scala#L25) but I wonder if it's too little or too much:

> Unless explicitly noted otherwise, containment and equality for key elements (i.e. set elements and map keys) refers to the `equals` / `hashCode` contract.

According to this you may use an equality that is incompatible with `equals` so you can't rely on the equality rules of a `Map` or `Set` that you didn't create. This is very useful for sorted maps and it is also required for normal maps with configurable equality (not supported by any of our standard implementations but this feature has been requested and you can do it with a custom implementation). Is there any principled reason why we shouldn't allow this?

Reply:
Couldn't this lead to some quite funky stuff like

```scala
val myWeirdMap: Map[K, V] = TreeMap(someWeirdOrdering)

myWeirdMap == myMap.to(ListMap) // false
myWeirdMap == myMap.to(HashMap) // false
myWeirdMap == myMap.to(SeqMap)
//... to(VectorMap) or to(TreeSeqMap) -->  false
```

The above will all at least be false because when building those maps, multiple key that are `==` will be squashed together so the maps won't even have the same size.

Reply:
Yes, that's a consequence of using different equalities. The response could either be "deal with it" or "require the same equality everywhere".

Reply:
@szeiger maybe we should then be removing or changing `MapView#+`

```scala
case class Foo(i: Int, j: Int) {
  override def equals(obj: Any): Boolean = obj match {
    case Foo(_i, _) => i == _i
    case _ => false
  }

  override def hashCode(): Int = i
}

// more granular than equality
implicit val ordering: Ordering[Foo] = 
  Ordering.by((foo: Foo) =>foo.i).orElseBy(_.j)

println((TreeMap(Foo(0, 0) -> 0) + (Foo(0, 1) -> 1)).get(Foo(0, 0)))
// Some(0)
println((TreeMap(Foo(0, 0) -> 0).view + (Foo(0, 1) -> 1)).get(Foo(0, 0)))
// Some(1)
```

Reply:
`+` is implemented in terms of `MapView.Concat`. We'd have to provide a dedicated implementation that uses the LHS equality. It's more difficult for `concat` itself which is not necessarily incorrect but has unintuitive semantics: You would expect it to use the LHS equality but the only efficient way to implement it is by using the RHS equality. We'd have to make equality first class in order to implement such operators efficiently and correctly. I suppose that's a good argument for not supporting it.

Reply:
After reflecting on it, I think the answer is clear.  An ordering defines an equality; `a == b` if and only if `a <= b` and `b <= a`; or equivalently, `a == b` if and only if `(a compare b) == 0`.

So default equality is irrelevant when you supply an ordering.  You ask the ordering, always.

However, if you _also_ use hash codes, then your hashes are ill-formed unless they agree with equality (i.e. `(a compare b) == 0` implies `a.## == b.##`.  Thus, you must (and it's the user's job to guarantee this) only use a hash-aware map with orderings that are consistent with hash equality.

Reply:
It sounds to me, then, like we should remove the overrides that make `MapView#concat` and `MapView#+` return `MapView`, and instead fall back to return `View[(K, V)]`.

Reply:
@joshlemer - That doesn't help much.  Two concatted maps with different ideas of equality are conceptually weird whether or not you type them as `MapView` or `View`.  Turning them into `Iterable` and joining them that way is even weirder; you have something which is nominally a map, but it doesn't act like one at all.

Reply:
`OrderAwareHashMap` already requires that hashing is consistent with the `Ordering`. It also requires a consistent `equals`. Since `hashCode` and `equals` are supposed to be consistent in general, you don't gain much by not requiring it in this specific case, and you get improved performance from using `equals` instead of `compare`. But that's for a single `Map` implementation.

`MapView.Concat` is the default implementation of `Map#concat` so we cannot remove it entirely but we can use `View.Concat` instead, which gives ues more consistent equality handling. `MapView.Concat` uses the RHS equality whereas `View.Concat` uses the LHS equality. The latter is IMHO the correct one because the result type (and therefore equality) is determined by the LHS, too. In other words: If `map1` and `map2` don't agree on equality, I expect `map1.concat(map2)` to give the same result as `map1.concat(map2.iterator)`.

I still think Josh is right and we have to remove `MapView.Concat` and make `MapView` concatenation return a plain `View` because there is no way to implement the desired semantics efficiently in `MapView.Concat`. You have to check keys from the RHS using the LHS equality. This is not possible in a lazy, view-like way. You'd have to materialize the concatenated map first. And without `MapView.Concat` we not only fix views but also `Map#concat` and `Map#+`.

We also have to remove the optimization in this PR or add some kind of marker to check whether a map or set uses the default equality.

Reply:
It's hard for me to imagine a use-case where you want to use an ordering more granular than equality, wouldn't a workaround be to simply create a new derived key from that key, which does expose a consistent ordering/equals implementation? For example, in the `Foo` example above, all the trouble could be avoided if they just keyed by `foo => (foo.i, foo.j)` (and if they want to keep the Foo value around, they could stick that in value, so...

```scala
TreeMap[(Int, Int), V]
//or if you need the Foo
TreeMap[(Int, Int), (Foo, V)]
```

however, if you're unconvinced by that, then yeah I agree we'll have to take out mapview concat and narrow this optimization to only occur when the argument is `HashMap | Map1 | Map2 | Map3 | Map4`

Reply:
@szeiger - I'm not sure I agree that `map1.concat(map2.iterator)` and `map1 union map2` (which is what `concat` effectively is for maps and sets) ought to give the same answer.  The counterargument is that this is also entirely reasonable for union: `(map1 concat map2).get(x)` is the same as `map2.get(x) orElse map1.get(x)`.

For views we can simply state that the latter is what we mean.  For non-views, we can choose differently.  Or maybe, if the disagreement is too weird, we want to introduce a separate `union` method that does the efficient and very sensible `map2.get(x) orElse map1.get(x)`.  I don't think we should just throw away the capability.  One of the worst aspects of 2.12-and-previous views was that they kept losing their specific types for no good apparent reason.  (Often there was a reason, but it was, as would be the case here, not terribly obvious.)

Reply:
> The counterargument is that this is also entirely reasonable for union: `(map1 concat map2).get(x)` is the same as `map2.get(x) orElse map1.get(x)`.

I don't see any way you could make this work without requiring the same equality everywhere. `map1.concat(map2)` has to pick one equality and a set of disjoint keys under that equality.

You could argue that it is nevertheless desirable for `MapView#concat`, but `MapView#+` is still weird and cannot be easily fixed. Even with a dedicated implementation (instead of reusing `MapView.Concat`) the best you can do for `get` is something like `lhs.get(key).or(if(key == appendedKey) Some(appendedValue) else None)`. Note that this uses the default equality ('==') because no other is available even though the `lhs` map may use a different equality.

Reply:
There isn't any strict requirement for the same equality at each stage--the map is sound if they're different--but you do run into weird cases when iterating the key-value pairs because you may get some partial repeats.  That is, you might get both `(k, v)` and `(c, u)` where `k == c` at least according to the left map (and maybe to default equality).  I'm not sure whether or not this is surprising enough to just avoid the issue by turning everything into a `View`.

However, `+` is easily fixed; just make maps expose--at least to `private[collection]`--their equality operation.  Then the view can defer to that.